### PR TITLE
Add CD+G Framework

### PR DESCRIPTION
**Pull request by mnadareski**
This PR adds most of the core functionality that will be needed for proper CD+G support in the future. There are some known issues with this code, such as requiring every subchannel to exist for this to work.

None of the code here is hooked up to any of the existing UI or player functionality and should not cause any issues for normal users.

Related issue: #8

### CODE REVIEW

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+                    {
+                        if(y - vOffsetTotal < 0)
+                            overflow[y] = this.DisplayData[x,y];
+                        else
+                            this.DisplayData[x, y - vOffsetTotal] = this.DisplayData[x,y];
+                    }
+
+                    // Fill in the now-empty pixels
+                    for(int i = 215; i > 215 - vOffsetTotal; i++)
+                    {
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: No single line statements with braces please

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+                        if(y + vOffsetTotal >= 216)
+                            overflow[(y + vOffsetTotal) % 216] = this.DisplayData[x,y];
+                        else
+                            this.DisplayData[x, y + vOffsetTotal] = this.DisplayData[x,y];
+                    }
+
+                    // Fill in the now-empty pixels
+                    for(int i = 0; i < vOffsetTotal; i++)
+                    {
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: No single line statements with braces please

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+                        if(x - hOffsetTotal < 0)
+                            overflow[x] = this.DisplayData[x,y];
+                        else
+                            this.DisplayData[x - hOffsetTotal, y] = this.DisplayData[x,y];
+                    }
+
+                    // Fill in the now-empty pixels
+                    for(int i = 299; i > 299 - hOffsetTotal; i++)
+                    {
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: No single line statements with braces please

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+                        if(x + hOffsetTotal >= 300)
+                            overflow[(x + hOffsetTotal) % 300] = this.DisplayData[x,y];
+                        else
+                            this.DisplayData[x + hOffsetTotal, y] = this.DisplayData[x,y];
+                    }
+
+                    // Fill in the now-empty pixels
+                    for(int i = 0; i < hOffsetTotal; i++)
+                    {
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: No single line statements with braces please

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+
+            for(int x = 0; x < 300; x++)
+            for(int y = 0; y < 6; y++)
+            {
+                this.DisplayData[x,y] = borderPreset.Color;
+            }
+
+            for(int x = 0; x < 300; x++)
+            for(int y = 210; y < 216; y++)
+            {
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: No single line statements with braces please

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+            for(int x = 297; x < 300; x++)
+            for(int y = 0; y < 216; y++)
+            {
+                this.DisplayData[x,y] = borderPreset.Color;
+            }
+
+            for(int x = 0; x < 300; x++)
+            for(int y = 0; y < 6; y++)
+            {
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: No single line statements with braces please

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+
+            for(int x = 0; x < 3; x++)
+            for(int y = 0; y < 216; y++)
+            {
+                this.DisplayData[x,y] = borderPreset.Color;
+            }
+
+            for(int x = 297; x < 300; x++)
+            for(int y = 0; y < 216; y++)
+            {
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: No single line statements with braces please

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+        /// on <see cref="DisplayData"/> for more details.
+        /// </remarks>
+        private void SetBorderColor(BorderPreset borderPreset)
+        {
+            if(borderPreset == null)
+                return;
+
+            for(int x = 0; x < 3; x++)
+            for(int y = 0; y < 216; y++)
+            {
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: No single line statements with braces please

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+            if(memPreset == null)
+                return;
+
+            // Skip in a consistent data case
+            if(consistentData && memPreset.Repeat > 0)
+                return;
+
+            for(int x = 3; x < 297; x++)
+            for(int y = 6; y < 210; y++)
+            {
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: No single line statements with braces please

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+        /// <summary>
+        /// Print the current color map to console
+        /// </summary>
+        public void DebugPrintScreen()
+        {
+            string screenDump = string.Empty;
+            for(int y = 0; y < 216; y++)
+            {
+                for(int x = 0; x < 300; x++)
+                {
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: No single line statements with braces please

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+            this.TransparentColor   = 0x00;
+        }
+
+        /// <summary>
+        /// Print the current color map to console
+        /// </summary>
+        public void DebugPrintScreen()
+        {
+            string screenDump = string.Empty;
+            for(int y = 0; y < 216; y++)
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs:72]: Separate statements with new line.

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+        {
+            if(!packet.IsCDGPacket())
+                return;
+
+            switch(packet.Instruction)
+            {
+                case SubchannelInstruction.MemoryPreset:
+                    var memoryPreset = new MemPreset(packet.Data);
+                    SetScreenColor(memoryPreset);
+                    break;
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs:95]: Separate exiting from statements with newline.

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+        private void SetScreenColor(MemPreset memPreset, bool consistentData = false)
+        {
+            if(memPreset == null)
+                return;
+
+            // Skip in a consistent data case
+            if(consistentData && memPreset.Repeat > 0)
+                return;
+
+            for(int x = 3; x < 297; x++)
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs:163]: Indentation

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+        /// The area that is considered the border is unclear. Please see the remarks
+        /// on <see cref="DisplayData"/> for more details.
+        /// </remarks>
+        private void SetBorderColor(BorderPreset borderPreset)
+        {
+            if(borderPreset == null)
+                return;
+
+            for(int x = 0; x < 3; x++)
+            for(int y = 0; y < 216; y++)
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: Indentation

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+            {
+                this.DisplayData[x,y] = borderPreset.Color;
+            }
+
+            for(int x = 297; x < 300; x++)
+            for(int y = 0; y < 216; y++)
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: Indentation

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+            {
+                this.DisplayData[x,y] = borderPreset.Color;
+            }
+
+            for(int x = 0; x < 300; x++)
+            for(int y = 0; y < 6; y++)
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: Indentation

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 17]
```diff
@@ -0,0 +1,367 @@
...
+            {
+                this.DisplayData[x,y] = borderPreset.Color;
+            }
+
+            for(int x = 0; x < 300; x++)
+            for(int y = 210; y < 216; y++)
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: Indentation

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 18]
```diff
@@ -0,0 +1,367 @@
...
+                pattern[i,1] = (byte)(b & (1 << 1));
+                pattern[i,2] = (byte)(b & (1 << 2));
+                pattern[i,3] = (byte)(b & (1 << 3));
+                pattern[i,4] = (byte)(b & (1 << 4));
+                pattern[i,5] = (byte)(b & (1 << 5));
+            }
+
+            // Now load the bitmap starting in the correct place
+            for(int x = 0; x < 12; x++)
+            for(int y = 0; y < 6; y++)
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: Indentation

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 19]
```diff
@@ -0,0 +1,367 @@
...
+            int vOffsetTotal = 12 + scroll.VScrollOffset;
+
+            // If we're scrolling horizontally
+            if(scroll.HScrollCommand == ScrollCommand.Positive
+                || (scroll.HScrollCommand == ScrollCommand.NoScroll && scroll.HScrollOffset > 0))
+            {
+                for(int y = 0; y < 216; y++)
+                {
+                    byte[] overflow = new byte[hOffsetTotal];
+                    for(int x = 299; x >= 0; x--)
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs:261]: Separate statements with newline

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 20]
```diff
@@ -0,0 +1,367 @@
...
+                        this.DisplayData[i,y] = copy ? overflow[i] : scroll.Color;
+                    }
+                }
+            }
+            else if(scroll.HScrollCommand == ScrollCommand.Negative)
+            {
+                for(int y = 0; y < 216; y++)
+                {
+                    byte[] overflow = new byte[hOffsetTotal];
+                    for(int x = 0; x < 300; x++)
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs:280]: Separate statementes with newline

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 21]
```diff
@@ -0,0 +1,367 @@
...
+        }
+
+        /// <summary>
+        /// Scroll the display according to the instruction
+        /// </summary>
+        /// <param name="scroll">Scroll with the new data</param>
+        /// <param name="copy">True if data wraps around on scroll, false if filled by a solid color</param>
+        private void ScrollDisplay(Scroll scroll, bool copy)
+        {
+            if(scroll == null)
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: What happens if scroll contains negatives due to errors?

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs [hunk 22]
```diff
@@ -0,0 +1,367 @@
...
+                        this.DisplayData[x,i] = copy ? overflow[i] : scroll.Color;
+                    }
+                }
+            }
+            else if(scroll.VScrollCommand == ScrollCommand.Negative)
+            {
+                for(int x = 0; x < 300; x++)
+                {
+                    byte[] overflow = new byte[vOffsetTotal];
+                    for(int y = 0; y < 216; y++)
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs:321]: Separate statements with newline

File: RedBookPlayer.Models/Hardware/Karaoke/LoadCLUT.cs [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+        /// <summary>
+        /// Interpret subchannel packet data as Load Color Lookup Table
+        /// </summary>
+        public LoadCLUT(byte[] bytes)
+        {
+            if(bytes == null || bytes.Length != 16)
+                return;
+
+            for(int i = 0; i < 8; i++)
+            {
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/LoadCLUT.cs]: No single line statements with braces please

File: RedBookPlayer.Models/Hardware/Karaoke/TileBlock.cs [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+            if(bytes == null || bytes.Length != 16)
+                return;
+
+            this.Color0 = (byte)(bytes[0] & 0x0F);
+            this.Color1 = (byte)(bytes[1] & 0x0F);
+            this.Row    = (byte)(bytes[2] & 0x1F);
+            this.Column = (byte)(bytes[3] & 0x3F);
+
+            for(int i = 0; i < 12; i++)
+            {
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Karaoke/TileBlock.cs]: No single line statements with braces please

File: RedBookPlayer.Models/Hardware/Player.cs [hunk 25]
```diff
@@ -1332,37 +1338,79 @@ private void SoundOutputStateChanged(object sender, PropertyChangedEventArgs e)
...
+            return parsedSubchannelData;
+        }
+
         /// <summary>
         /// Reformat raw subchannel data for multiple sectors
         /// </summary>
         /// <param name="subchannelData">Raw subchannel data to format</param>
         /// <returns>Dictionary mapping subchannel to formatted data</returns>
-        public Dictionary<char, byte[]> ConvertSubchannels(byte[] subchannelData)
+        private Dictionary<char, byte[]> ConvertSubchannels(byte[] subchannelData)
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Player.cs]: private modifier is implicit

File: RedBookPlayer.Models/Hardware/Player.cs [hunk 26]
```diff
@@ -1332,37 +1338,79 @@ private void SoundOutputStateChanged(object sender, PropertyChangedEventArgs e)
...
         {
             if(subchannelData == null || subchannelData.Length % 96 != 0)
                 return null;
 
+            // Parse the subchannel data, if possible
+            var parsedSubchannelData = ParseSubchannels(subchannelData);
+            return ConvertSubchannels(parsedSubchannelData);
```
Comment (claunia) [inline RedBookPlayer.Models/Hardware/Player.cs:1380]: Separate return statements with newline

**Review by claunia — CHANGES_REQUESTED** [review]
Are you using the editorconfig file for the suite? Most of the problems are formatting ones that should've been fixed by it.

File: RedBookPlayer.Models/Hardware/Player.cs (hunk 25 above)
Comment (mnadareski) [reply RedBookPlayer.Models/Hardware/Player.cs]: I can provide additional cleanup of this in a later PR. The use of `private` is across a lot of places.

**Comment by mnadareski** [comment]
I thought that the editor config file was in here, since it caught things like the `for (` issues previously.

File: RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs (hunk 21 above)
Comment (mnadareski) [reply RedBookPlayer.Models/Hardware/Karaoke/KaraokeDisplay.cs]: I can add the secondary checks for negatives here.

**Review by claunia — APPROVED** [review]

**Pull request merged**